AzureAD/microsoft-authentication-library-common-for-android
# Add base64 encode for the state parameter

## The Problem
When generating the `state` parameter in Android, we are appending the `taskId` which is an Integer and two UUIDs as follows
```java
String.format("%d%s%s", taskId, SPLITTER, uuids)
```

The generated state is then passed to the `AuthorizationRequest` which builds an authorization URL as a request parameter. Once the authorization is complete, we compare the returned state parameter with the parameter generated on the SDK.

This works fine for most device languages. 
But when the device language is set to "Ar", there is a difference in the returned state parameter and the generated parameter.

The `requestStateParameter` is the one generated by the SDK, the `state` is returned from the authorization redirect.

![state](https://user-images.githubusercontent.com/25629064/170461388-c86eba35-1514-4b3f-a15a-7b6e60168c5b.PNG)

## Why is this happening?
The `String.format` above that generates the state parameter makes use of the device's Locale. This means that the `taskId` will not make use of ASCII characters in some cases e.g. when the device's language is "ar". 

## Changes Made
1. We need to ensure we use a single Locale when generating the state parameter in Android. 
2. It may be hard for us to keep track of which state generators create a URL-unsafe state parameter, hence we will Base64 encode the state parameter value before sending it in the Authorization request. This ensures that we always have a state parameter that is URL safe.

### Reproducing the bug.
1. Change your device language to "AR"
2. Checkout to dev branch and run MSAL test app
3. Perform interactive login with any account
4. You will get this error "Returned state from authorize endpoint is not the same as the one sent"

https://user-images.githubusercontent.com/25629064/170463766-d2be8a16-e391-4341-92c6-6666342a2e6e.mp4

## Testing
1. Checkout to the branch `git/state-mismatch-exception`
2. Change your device language to "AR"
3. Run MSAL test app and ensure broker is not installed in the device.
4. Perform interactive login with any account
